### PR TITLE
[BEAM-8988] RangeTracker for _CustomBigQuerySource

### PR DESCRIPTION
**Pull request by kamilwu**
Also, the PR effectively fixes https://issues.apache.org/jira/browse/BEAM-8528

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [x] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [x] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/)
Java | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_SparkStructuredStreaming/lastCompletedBuild/)
Python | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python2/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python35/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/) | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python2_PVR_Flink_Cron/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python35_VR_Flink/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/)
XLang | --- | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/) | --- | --- | ---

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website
--- | --- | --- | --- | ---
Non-portable | [Build Status](https://builds.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PreCommit_PythonLint_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/) 
Portable | --- | [Build Status](https://builds.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

### CODE REVIEW

**Comment by kamilwu** [comment]
Run Python 3.7 PostCommit

**Comment by kamilwu** [comment]
R: @robertwb  I'd be happy if you could take a took.

**Comment by kamilwu** [comment]
Run Python 2 PostCommit

**Comment by pabloem** [comment]
Run Python 3.7 PostCommit

**Comment by pabloem** [comment]
image
Looking. It seems that precommit Failed due to warnings?

**Comment by udim** [comment]
Errors appear before warnings. `FnApiLogRecordHandlerTest.test_context` failed

File: sdks/python/test-suites/portable/py2/build.gradle [hunk 1]
```diff
@@ -147,10 +148,10 @@ task crossLanguagePortableWordCount {
   }
 }
 
-task postCommitIT {
+task postCommitPy2IT {
   dependsOn 'installGcpTest'
   dependsOn 'setupVirtualenv'
-  dependsOn ':runners:flink:1.8:job-server:shadowJar'
```
Comment (pabloem) [inline sdks/python/test-suites/portable/py2/build.gradle]: I don't know much about these tests. Why did you upgrade this?

**Comment by pabloem** [comment]
Run Python PreCommit

File: sdks/python/test-suites/portable/py2/build.gradle (hunk 1 above)
Comment (kamilwu) [reply sdks/python/test-suites/portable/py2/build.gradle]: This is a task that executes BigQuery read IT tests on Flink. It was added by me in the previous PR. I upgraded this to keep it consistent with Py 3.x versions:

https://github.com/apache/beam/blob/166ed817ab11f1d92ab2b6e0127673be5d0ea9c3/sdks/python/test-suites/portable/py37/build.gradle#L47-L50

Also, I included this task in the PostCommit suite (I accidentally hadn't done this earlier)

**Comment by kamilwu** [comment]
> Errors appear before warnings. FnApiLogRecordHandlerTest.test_context failed.

Do we have a ticket for this? I've seen this test fails quite frequently in PreCommit Cron (but not always)

**Comment by kamilwu** [comment]
Adding a commit that fixes https://issues.apache.org/jira/browse/BEAM-8995

**Comment by kamilwu** [comment]
Run Python 2.7 PostCommit

**Comment by kamilwu** [comment]
Run Python 3.7 PostCommit

**Comment by kamilwu** [comment]
Run Python 2 PostCommit

**Comment by kamilwu** [comment]
Run Python 3.6 PostCommit

**Comment by kamilwu** [comment]
Run Python 3.5 PostCommit

**Comment by kamilwu** [comment]
Run Python 3.7 PostCommit

**Comment by kamilwu** [comment]
Run Python 3.5 PostCommit

**Comment by kamilwu** [comment]
Run Python 2 PostCommit

**Comment by kamilwu** [comment]
Run Python 3.5 PostCommit

**Comment by kamilwu** [comment]
Run Python 3.6 PostCommit

**Comment by kamilwu** [comment]
One check left.

**Comment by tvalentyn** [comment]
> Do we have a ticket for this? I've seen this test fails quite frequently in PreCommit Cron (but not always)
Yes: BEAM-8974.

File: sdks/python/apache_beam/io/gcp/bigquery.py [hunk 2]
```diff
@@ -554,6 +555,8 @@ def _convert_to_tuple(cls, table_field_schemas):
             for x in table_field_schemas]
 
   def decode(self, value):
+    if isinstance(value, bytes):
```
Comment (tvalentyn) [inline sdks/python/apache_beam/io/gcp/bigquery.py]: Is the value here always `bytes`? If so, do we need the if? In general it is strongly discouraged to have apis that can take both bytes and strings.

Comment (TheNeuralBit) [reply sdks/python/apache_beam/io/gcp/bigquery.py]: Unless @kamilwu is available to resolve this soon, I'm inclined to just file a folow-up jira and merge this as is to get postcommits green.

Comment (kamilwu) [reply sdks/python/apache_beam/io/gcp/bigquery.py]: Indeed, it's unnecessary. Pushing a fix

**Comment by kamilwu** [comment]
Run Python 2 PostCommit

**Comment by kamilwu** [comment]
Run Python 3.5 PostCommit

**Comment by kamilwu** [comment]
Run Python PreCommit

**Comment by kamilwu** [comment]
Run Python 2 PostCommit

**Comment by kamilwu** [comment]
Run Python 3.5 PostCommit

**Comment by kamilwu** [comment]
Run Python PreCommit

**Comment by kamilwu** [comment]
Run Python 3.7 PostCommit

**Comment by kamilwu** [comment]
Run Python 3.5 PostCommit

**Comment by tvalentyn** [comment]
Run Python PreCommit

**Comment by TheNeuralBit** [comment]
Run Python PreCommit

**Comment by kamilwu** [comment]
Run Python 3.5 PostCommit

**Comment by kamilwu** [comment]
Run Python 2 PostCommit

**Comment by kamilwu** [comment]
Run Python 2 PostCommit

**Comment by kamilwu** [comment]
Run Python 3.5 PostCommit

**Pull request merged**